### PR TITLE
Make ocs optional

### PR DESCRIPTION
**Pull request by jakub-dzon**
This PR introduces support for `OBC_AUTO_CREATE` ENV variable, which controls whether OBCs should be automatically created for EdgeDevices.
`make run` automatically disables OBC auto-creation.

### CODE REVIEW

File: controllers/edgedevice_controller.go [hunk 1]
```diff
@@ -69,33 +71,48 @@ func (r *EdgeDeviceReconciler) Reconcile(ctx context.Context, req ctrl.Request)
...
-			if errors.IsNotFound(err) {
-				logger.Info("Failed to find an existing OBC for the device. Creating new OBC", "edgeDevice", edgeDevice)
-				obc, err = r.Claimer.CreateClaim(ctx, edgeDevice)
-				if err != nil {
-					logger.Error(err, "Cannot create object bucket claim for the device", "EdgeDevice Name", edgeDevice.Name, "EdgeDevice Namespace", edgeDevice.Namespace)
-					return ctrl.Result{Requeue: true}, err
-				}
-			} else {
-				logger.Error(err, "Failed to get OBC for the device", "edgeDevice", edgeDevice)
+	if r.ObcAutoCreate {
```
[the remark this answers is not in the document]
Comment (masayag) [reply controllers/edgedevice_controller.go:74]: Also, what will be the behavior when changing the value in the config map ? the deployment has to be restarted in order to reflect the new value ?

**Review by masayag — COMMENTED** [review]
Did you get a chance to verify it on top of minikube ?
No need to update the roles not to list to [obc resources](https://github.com/jakub-dzon/k4e-operator/blob/main/config/rbac/role.yaml#L78) which aren't exist on minikube ? Perhaps we need to maintain also operator.yaml clusters without OBC.

File: controllers/edgedevice_controller.go [hunk 2]
```diff
@@ -69,33 +71,48 @@ func (r *EdgeDeviceReconciler) Reconcile(ctx context.Context, req ctrl.Request)
...
+				return ctrl.Result{Requeue: true, RequeueAfter: time.Second * 5}, err
+			}
 		}
 	}
 	return ctrl.Result{}, nil
 }
 
+func (r *EdgeDeviceReconciler) createOrGetObc(ctx context.Context, edgeDevice *managementv1alpha1.EdgeDevice) (*obv1.ObjectBucketClaim, error) {
+	obc, err := r.Claimer.GetClaim(ctx, edgeDevice.Name, edgeDevice.Namespace)
+	if err != nil {
```
Comment (eloycoto) [inline controllers/edgedevice_controller.go:93]: I would do something like 

```
if err == nil {
logger.Error(err, "Failed to get OBC for the device", "edgeDevice", edgeDevice)
			return nil, err
}
logger := log.FromContext(ctx)
if errors.IsNotFound(err) {
```

It's a bit clear what is the normal flow function.

File: controllers/edgedevice_controller_test.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+		err = k8sClient.Get(ctx, client.ObjectKey{Namespace: ed.GetNamespace(), Name: *ed.Status.DataOBC}, &obc)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(obc.Spec.StorageClassName).To(BeEquivalentTo("openshift-storage.noobaa.io"))
+	})
+})
+
+func getExpectedEdgeDevice(ctx context.Context, objectKey client.ObjectKey) v1alpha1.EdgeDevice {
+	var ed v1alpha1.EdgeDevice
+	err := k8sClient.Get(ctx, objectKey, &ed)
+	Expect(err).ToNot(HaveOccurred())
```
Comment (eloycoto) [inline controllers/edgedevice_controller_test.go:95]: I would do a `ExpectWithOffset` so you can know where it fails.

**Review by eloycoto — COMMENTED** [review]
Thanks for doing the tests, a few minor comments!

File: controllers/edgedevice_controller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jakub-dzon) [reply controllers/edgedevice_controller.go:74]: Yes, during my testing change in the config map haven't triggered re-deploy.

[the remark this answers is not in the document]
Comment (jakub-dzon) [reply controllers/edgedevice_controller.go:74]: Yeah, it looks nicer without nesting. I am not so sure it's helping future development - if there is any new code common for all cases, the condition has to be inverted back.

File: controllers/edgedevice_controller_test.go (hunk 3 above)
Comment (jakub-dzon) [reply controllers/edgedevice_controller_test.go:95]: Thanks for the hint; I didn't know there is something like that.

**Review by eloycoto — APPROVED** [review]

**Comment by machacekondra** [comment]
> > Did you get a chance to verify it on top of minikube ? No need to update the roles not to list to [obc resources](https://github.com/jakub-dzon/k4e-operator/blob/main/config/rbac/role.yaml#L78) which aren't exist on minikube ? Perhaps we need to maintain also operator.yaml clusters without OBC.
> 
> Yes, I've tested it on Minikube with flag set to false and in OCP with flag in both settings. Existing manifests work fine and roles that refer to non-existing resources are not causing issues.

@jakub-dzon As a newcomer to project I've just took minikube and try to setup the project. I wasn't aware of must have OCS. But when troubleshooting issues I've saw that we could just have [configurable](https://github.com/jakub-dzon/k4e-operator/blob/main/internal/storage/storage.go#L61) `bucketHost` `caBundle` and we could have working data sync also on minikube. Is there any reason we don't do so?

**Comment by jakub-dzon** [comment]
@machacekondra - @ydayagi is currently working on making all the s3 endpoint details configurable (https://issues.redhat.com/browse/ECOPROJECT-272). At the moment you can either turn the data sync off using `OBC_AUTO_CREATE` env var or use OCS with automatically-created OBCs.

**Pull request merged**